### ISSUE TITLE
Remote query execution design considerations

### ISSUE THREAD

**Issue by kreuzerkrieg** [yandex/ClickHouse#5662]
Hi,
I want to play around with remote query execution on CH cluster setup, something like `Distributed` table engine but without the part when something executed locally and then sent to other shards from initiating machine. Sound like something simple to implement, just create another table engine that uses `Connection` to send queries to remote members of `Cluster` and you are done.

Well, since nothing produced and sent to remotes `IBlockOutputStream` (for `BlockIO`'s `out` member ) should be implemented and no `IBlockInputStream` should be used, something like one could observe in `InterpreterInsertQuery` when `insert` comes without `select`. However the problem is that query initiating side (the client) want to report progress of each execution on each cluster node. I dont know how it could be done except using `IBlockInputStream` to provide the `in` for `BlockIO` which is returned by `IInterpreter`'s `execute` even when `in` technically is not needed.

Long story short, by hacking, twisting things and breaking interfaces I achieved what I wanted, but it is quite ugly. Is there something I'm not aware of that could help me to achieve the above without committing crimes against humanity?

**Comment by filimonov** [comment]
Aren't you trying to reinvent the `remote` table function?

See docs: https://clickhouse.yandex/docs/en/query_language/table_functions/remote/

And samples in tests: https://github.com/yandex/ClickHouse/search?l=SQL&q=remote

**Comment by kreuzerkrieg** [comment]
Hm... it works on top of `Distributed` which works with `DistributedBlockOutputStream` and (I guess) `DirectoryMonitor` which in turn somewhat confusing for me at this moment, but I will dig into the code and check it right away.
Now the question, what if table function is not good enough for me, and I want to hide all function arguments and configurations behind table engine...

**Comment by alexey-milovidov** [comment]
@filimonov 
`remote` table function works exactly as Distributed engine, because it creates temporary Distributed table to process the query.

@kreuzerkrieg 
> but without the part when something executed locally and then sent to other shards

The intention is not quite clear...
Do you want to:
- execute a quert on shards to complete stage and union results (without proper merging of them)?
(you can get it with `distributed_group_by_no_merge` setting)
- execute a query on a single shard to complete stage?
(this is what Distributed engine does automatically when there is only single shard)
- fetch source data from shards and do all the required processing on initiating server?
(this is rather slow due to high network usage, but you can easily achieve it in code by changing requesting `QueryProcessingStage::WithMergeableState` to `QueryProcessingStage::FetchColumns`)

**Comment by kreuzerkrieg** [comment]
@alexey-milovidov 
I think none of above... I will provide an example and correct me if I wrong

I have something like:
```
INSERT INTO T2
SELECT * FROM T1
```
Where T1 is, lets say, a real MergeTree table on each cluster node, and T2 is created using some kind of table engine which in turn looks like following:
```
CREATE TABLE T2
(
    `Foo` UInt32
)
ENGINE = SomeEsotericEngine(cluster, default, some_real_table)
```
then, when the `insert` query is executed it distributed to run on each node in such a fashion that it will be translated into a query
```
INSERT INTO some_real_table
SELECT * FROM T1
```
which just will take all data from local table `T1` and insert it into local table `some_real_table`, repeat it on each node, no cross node merges/synchronizations/younameit of data from each of above tables.

The only thing I expect to be reflected on initiating side is a combined progress of all queries shown

**Comment by filimonov** [comment]
Just to understand the problem better - it seems like you still need to do one query per server, so why don't you execute the query directly on needed server?

**Comment by kreuzerkrieg** [comment]
why execute 150 queries on 150 servers instead to let the machine do that?
And I still want to see progress of all 150 queries combined. In one client.
